### PR TITLE
Added document links for xsi:schemaLocation

### LINKED ISSUE #666: Add support for `textDocument/documentLink` for xsi:schemaLocation
Given this XML Schema:

```xml
<?xml version="1.0"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
targetNamespace="https://www.w3schools.com"
xmlns="https://www.w3schools.com"
elementFormDefault="qualified">

<xs:element name="note">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="to" type="xs:string"/>
      <xs:element name="from" type="xs:string"/>
      <xs:element name="heading" type="xs:string"/>
      <xs:element name="body" type="xs:string"/>
    </xs:sequence>
  </xs:complexType>
</xs:element>

</xs:schema>
```

and this XML:

```xml
<note xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	  xmlns="https://www.w3schools.com"
	  xsi:schemaLocation="https://www.w3schools.com note.xsd">
	<to></to>
	<from></from>
	<heading></heading>
	<body></body>
</note>
```

The `note.xsd` from xsi:schemaLocation="https://www.w3schools.com note.xsd" must be a documentLink to open it it with Ctrl+Click. 

Pay attention that xsi:schemaLocation can contain several xsd .

### PR DESCRIPTION
**Pull request by datho7561**
Every second token for the value of xsi:schemaLocation is turned into a
document link to the referenced schema file.

Closes #666

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 1]
```diff
@@ -12,40 +12,62 @@
...
+			// to highlight the end, and one to make up for "
+			int end = valNode.getStart() + locPairMatcher.end(2) + 2;
+			schemaLocationValuePairs.put(namespaceURI,
+					new SchemaLocationHint(start, end, locationHint, this));
+		}
 	}
 
 	public String getLocationHint(String namespaceURI) {
-		return schemaLocationValuePairs.get(namespaceURI);
+		return schemaLocationValuePairs.get(namespaceURI).getHint();
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:57]: You must prevent from NPE because namespaceURI could not be defined in the xsi:schemaLocation. Do liek this:

```java
public String getLocationHint(String namespaceURI) {
	SchemaLocationHint schemaLocationHint = schemaLocationValuePairs.get(namespaceURI); 
	return schemaLocationHint != null ? schemaLocationHint.getHint() : null;
}
```

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 2]
```diff
@@ -12,40 +12,62 @@
...
+			int start = valNode.getStart() + locPairMatcher.start(2);
+			// matcher matches end correctly, but range needs to be one past end
+			// to highlight the end, and one to make up for "
+			int end = valNode.getStart() + locPairMatcher.end(2) + 2;
+			schemaLocationValuePairs.put(namespaceURI,
+					new SchemaLocationHint(start, end, locationHint, this));
+		}
 	}
 
 	public String getLocationHint(String namespaceURI) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:56]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java (hunk 1 above)
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:57]: Thank you!

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 3]
```diff
@@ -12,40 +12,62 @@
...
-	private final Map<String, String> schemaLocationValuePairs;
+	private final Map<String, SchemaLocationHint> schemaLocationValuePairs;
 
 	private final DOMAttr attr;
 
+	// The text to match is of the form:
+	// http://example.org root.xsd http://example.org bison.xsd http://example.org potato.xsd
+	private static final Pattern SCHEMA_LOCATION_PAIR_PATTERN = Pattern.compile("\\s*([^\\s]+)\\s+([^\\s]+)\\s*");
+
 	public SchemaLocation(String base, DOMAttr attr) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:35]: Remove the base parameter which is never used.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 4]
```diff
@@ -12,40 +12,72 @@
...
 public class SchemaLocation {
 
-	private final Map<String, String> schemaLocationValuePairs;
+	private final Map<String, SchemaLocationHint> schemaLocationValuePairs;
 
 	private final DOMAttr attr;
 
-	public SchemaLocation(String base, DOMAttr attr) {
+	// The text to match is of the form:
+	// http://example.org root.xsd http://example.org bison.xsd http://example.org potato.xsd
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:32]: This is not a valid `schemaLocation` since the namespace URI are the same (same namespace is not allowed)

Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java]: Okay. I resolved this and put in a proper example.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/XMLSchemaDocumentLinkTest.java [hunk 5]
```diff
@@ -35,4 +35,58 @@ public void noNamespaceSchemaLocation() throws BadLocationException {
...
+
+	public void blankSchemaLocation() throws BadLocationException {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" + //
+		"<beans xmlns=\"http://www.springframework.org/schema/beans\"\r\n" + //
+		"       xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n" + //
+		"       xmlns:camel=\"http://camel.apache.org/schema/spring\"\r\n" + //
+		"       xmlns:cxf=\"http://camel.apache.org/schema/cxf\"\r\n" + //
+		"       xsi:schemaLocation=\"\" />";
+		XMLAssert.testDocumentLinkFor(xml, "src/test/resources/xsd/bean.xml");
+
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/XMLSchemaDocumentLinkTest.java:90]: Add 2 tests  with bad schemaLocation to check that there are none problem with NPE other things : 

 * with 1 token : xsi:schemaLocation=\"http://www.example.org/schema/beans\"
 * with 3 tokens: xsi:schemaLocation=\"http://www.example.org/schema/beans xsd/spring-beans.xsd http://www.example.org/schema/spring xsd/camel-spring.xsd\"

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 6]
```diff
@@ -12,40 +12,72 @@
...
-		StringTokenizer st = new StringTokenizer(value);
-		do {
-			String namespaceURI = st.hasMoreTokens() ? st.nextToken() : null;
-			String locationHint = st.hasMoreTokens() ? st.nextToken() : null;
+		Matcher locPairMatcher = SCHEMA_LOCATION_PAIR_PATTERN.matcher(value);
+		while (locPairMatcher.find()) {
+			String namespaceURI = locPairMatcher.group(1);
+			String locationHint = locPairMatcher.group(2);
 			if (namespaceURI == null || locationHint == null)
 				break;
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:44]: Could we add braces for the if statement?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 7]
```diff
@@ -12,40 +12,72 @@
...
+					new SchemaLocationHint(start, end, locationHint, this));
+		}
 	}
 
+	/**
+	 * Get the location hint, as a string, that is associated with the given namespace URI.
+	 * 
+	 * If the given namespace URI was not referred to in this xsi:schemaLocation, then null is returned.
+	 * 
+	 * @param namespaceURI The namespace URI to fid the location hint for
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:61]: ... to find ...

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 8]
```diff
@@ -12,40 +12,72 @@
...
+	 * @return The associated location hint, as a string, or null
+	 * if the namespace was not refered to in xsi:schemaLocation
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:63]: ... not referred to ...

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java [hunk 9]
```diff
@@ -12,40 +12,72 @@
...
+		SchemaLocationHint locHint = schemaLocationValuePairs.get(namespaceURI);
+		return locHint == null ? null : locHint.getHint();
 	}
 
 	public DOMAttr getAttr() {
 		return attr;
 	}
 
+	/**
+	 * Gets all the location hints given in this xsi:schemaLocation attribute
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocation.java:75]: Returns all the ...

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (c) 2018 Red Hat Inc.
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java:2]: Is this a new class you created yourself? If yes, should be 2020

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+	private final SchemaLocation parent;
+
+	/**
+	 * Create a new SchemaLocationHint object that represents one of the location hints
+	 * given in the parent <code>SchemaLocation</code>.
+	 *
+	 * @param start The offset from the beginning of the document where this location hint starts
+	 * @param end The offset from the beginning of the document where this location hint ends
+	 * @param hint The hint to the location of a schema (A URI that points to a schema)
+	 * @param parent The <coed>SchemaLocation</code> in which this hint was given
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java:33]: Please align the parameters like this: https://github.com/eclipse/lemminx/blob/022615e0c9fb2cf7af12b1ad29a16380a2768a04/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/StringUtils.java#L434-L444 thank you

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+	 */
+	public SchemaLocationHint(int start, int end, String hint, SchemaLocation parent) {
+		this.start = start;
+		this.end = end;
+		this.hint = hint;
+		this.parent = parent;
+	}
+
+	/**
+	 * Get the location hint that this SchemaLocationHint represents
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/SchemaLocationHint.java:43]: Returns

**Comment by angelozerr** [comment]
+1 to merge it. @xorye if you are OK please do that.
